src/ml_tooling/data/sql.py

```diff
@@ -12,13 +21,94 @@ class SQLDataset(Dataset, metaclass=abc.ABCMeta):
     SQLDataset takes a SQLAlchemy connection object or string URI to create the engine
     """
 
-    def __init__(self, conn, **engine_kwargs):
+    table: Optional[sa.Table] = None
+
+    def __init__(self, conn: Connectable, schema: str, **kwargs):
         if isinstance(conn, Connectable):
             self.engine = conn
         elif isinstance(conn, str):
-            self.engine = sa.create_engine(conn, **engine_kwargs)
+            self.engine = sa.create_engine(conn, **kwargs)
         else:
             raise ValueError(f"Invalid connection")
+        if self.table is not None and self.table.schema is not None:
+            raise DatasetError(
+                f"{self.table.schema.name} cannot have a defined schema - "
+                f"remove the schema declaration"
+            )
+        self.schema = schema
+
+    @contextmanager
+    def create_connection(self):
+        conn = self.engine.connect().execution_options(
+            schema_translate_map={None: self.schema}
+        )
+        try:
+            yield conn
+        finally:
+            conn.close()
+
+    def _load_training_data(self, *args, **kwargs) -> Tuple[pd.DataFrame, DataType]:
+        with self.create_connection() as conn:
+            return self.load_training_data(conn)
+
+    def _load_prediction_data(self, *args, **kwargs) -> pd.DataFrame:
+        with self.create_connection() as conn:
+            return self.load_prediction_data(conn)
+
+    def _dump_data(self):
+        logger.info(f"Dumping data from {self.table}")
+        logger.debug(f"Dumping data from {self.engine}/{self.table.name}")
+        stmt = sa.select([self.table])
+        try:
+            with self.create_connection() as conn:
+                data = conn.execute(stmt).fetchall()
+        except DBAPIError:
+            logger.exception("Data dump failed")
+            raise
+
+        logger.info("Data dumped...")
+        return data
+
+    def _setup_table(self, conn: sa.engine.Connection):
+        if not self.engine.dialect.has_schema(self.engine, self.schema):
+            self.engine.execute(sa.schema.CreateSchema(self.schema))
+        self.table.drop(bind=conn, checkfirst=True)
+        self.table.create(bind=conn)
+
+    def _insert_data(self, data):
+        logger.info(f"Inserting data into {self.table.name}")
+        with self.create_connection() as conn:
+            trans = conn.begin()
+            try:
+                self._setup_table(conn)
+                insert = self.table.insert()
+                conn.execute(insert, data)
+                trans.commit()
+            except DBAPIError:
+                logger.exception("Insert data failed")
+                trans.rollback()
+                raise
+
+    def copy_to(self, target: "SQLDataset") -> "SQLDataset":
```

Comment:
We should add these methods to the ABC class as well.

Off the top of my head we should implement `_insert_data`, `_dump_data` and `copy_to`, though maybe the naming should be more generalized - these should also apply to FileDataset, so we can dump a SQLDataset to a FileDataset and vice versa

Reply:
Also means that the typehint should be a `Dataset`

Reply:
I have added an issue for this #396. Can we merge this as we need it for refactoring, and do these later?

Reply:
👍